### DISCUSSION TITLE
Questions about the behavior of STSAssumeRoleSessionCredentialsProvider

### DISCUSSION THREAD

**Discussion by isaac33zhang** [aws/aws-sdk-java#2913] [Q&A]
Hi
I am trying to use STSAssumeRoleSessionCredentialsProvider with AWS KMS client. As is known, STSAssumeRoleSessionCredentialsProvider spins up a background thread to refresh the credentials. We have a couple of questions about it:
1. When does the background thread exit?
For example, for a single request, if the workflow is just we create a credential provider, create a kms client with it, and do a decrypt operation, does the background thread get closed after the workflow is completed?
2. If we put the instantiated credential provider object in a cache, will the backgroud thread of it keep running?
3. If the answer to question 2 is yes, will the background thread exit if the object gets evicted from the cache?
Thanks

**Comment by debora-ito** [comment]
1. The background thread will only shutdown when you call `close()` on the STSAssumeRoleSessionCredentialsProvider when the credential provider is no longer needed. This is mentioned in the docs - https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/STSAssumeRoleSessionCredentialsProvider.html
2. Yes
3. No, you need to explicitly close the credential provider.
